pkg/didcomm/crypto/authcrypt/encrypt.go

```diff
@@ -0,0 +1,288 @@
+/*
+Copyright SecureKey Technologies Inc. All Rights Reserved.
+
+SPDX-License-Identifier: Apache-2.0
+*/
+
+package authcrypt
+
+import (
+	"crypto/cipher"
+	"crypto/ecdsa"
+	"crypto/elliptic"
+	"crypto/rand"
+	"crypto/sha256"
+	"encoding/base64"
+	"encoding/json"
+	"fmt"
+	"reflect"
+	"sort"
+	"strings"
+
+	josecipher "github.com/square/go-jose/v3/cipher"
+	errors "golang.org/x/xerrors"
+
+	"github.com/btcsuite/btcutil/base58"
+	"golang.org/x/crypto/chacha20poly1305"
+	"golang.org/x/crypto/poly1305"
+)
+
+// This package deals with Authcrypt encryption for Packing/Unpacking DID Comm exchange
+// Chacha20Poly35 encryption/authentication
+
+// RandReader is a cryptographically secure random number generator (stubbed out in tests).
+var RandReader = rand.Reader
+
+// Crypter represents an Authcrypt Encrypter (Decrypter) that outputs/reads JWE envelopes
+type Crypter struct {
+	Sender     interface{}
+	Recipients []interface{}
+	Alg        string
+	NonceSize  int
+}
+
+// JweEnvelope represents a JWE envelope as per the Aries Encryption envelop specs
+type JweEnvelope struct {
+	Protected  jweHeaders  `json:"Protected,omitempty"`
+	Recipients []Recipient `json:"recipients,omitempty"`
+	Aad        string      `json:"aad,omitempty"`
+	Iv         string      `json:"iv,omitempty"`
+	Tag        string      `json:"tag,omitempty"`
+	CipherText string      `json:"ciphertext,omitempty"`
+}
+
+// jweHeaders are the Protected JWE headers in a map format
+type jweHeaders map[string]string
+
+// Recipient is a recipient of an envelop including the shared encryption key
+type Recipient struct {
+	EncryptedKey string           `json:"encrypted_key,omitempty"`
+	Header       RecipientHeaders `json:"header,omitempty"`
+}
+
+// RecipientHeaders are the recipient headers
+type RecipientHeaders struct {
+	Apu string `json:"apu,omitempty"`
+	Iv  string `json:"iv,omitempty"`
+	Tag string `json:"tag,omitempty"`
+	Kid string `json:"kid,omitempty"`
+	Oid string `json:"oid,omitempty"`
+}
+
+// NewAuthCrypter will create an encrypter instance to 'AuthCrypt' payloads for the given sender and recipients arguments
+// and the encryption alg argument. Possible algorithms supported are:
+// C20P (chacha20-poly1035 ietf)
+// XC20P (xchacha20-poly1035 ietf)
+func NewAuthCrypter(sender interface{}, recipients []interface{}, alg string) (*Crypter, error) {
+	var nonceSize int
+	switch alg {
+	case "C20P":
+		nonceSize = chacha20poly1305.NonceSize
+	case "XC20P":
+		nonceSize = chacha20poly1305.NonceSizeX
+	default:
+		return nil, errors.New(fmt.Sprintf("Alg %s not supported", alg))
+	}
+
+	c := &Crypter{
+		sender,
+		recipients,
+		alg,
+		nonceSize,
+	}
+
+	typeofS := reflect.TypeOf(sender)
+	if typeofS != reflect.TypeOf(&ecdsa.PrivateKey{}) {
```

Comment:
We'll need support for `ed25519.PrivateKey` as well, see #96

Reply:
yes, eventually, not sure if we can derive ECDH keys from ed25519 keys though.. this needs to be investigated ..

Reply:
for now, we're supporting keys used by Chacha20Poly1035 and Salsa20, these are 32 byte keys.